python/lsst/sims/featureScheduler/basis_functions.py

```diff
@@ -732,6 +804,147 @@ def __call__(self, **kwargs):
+    """Remove the bonus for staying in the same filter if certain conditions are met.
+
+    If the moon rises/sets or twilight starts/ends, it makes a lot of sense to consider
+    a filter change. This basis function rewards if it matches the current filter, the moon rises or sets,
+    twilight starts or stops, or there has been a large gap since the last observation.
+
+    """
+
+    def __init__(self, survey_features=None, condition_features=None, time_lag_min=10., time_lag_max=30.,
+                 time_lag_boost=60., boost_gain=2.0, unseen_before_lag=False,
+                 filtername='r', tag=None, twi_change=-18., proportion=1.0, aways_available=False):
+        """
+        Paramters
+        ---------
+        time_lag : float (10.)
+            If there is a gap between observations longer than this, let the filter change (minutes)
+        twi_change : float (-18.)
+            The sun altitude to consider twilight starting/ending
+        """
+        self.time_lag_min = time_lag_min / 60. / 24.  # Convert to days
+        self.time_lag_max = time_lag_max / 60. / 24.  # Convert to days
+        self.time_lag_boost = time_lag_boost / 60. / 24.
+        self.boost_gain = boost_gain
+        self.unseen_before_lag = unseen_before_lag
+
+        self.twi_change = np.radians(twi_change)
+        self.filtername = filtername
+        self.proportion = proportion
+        self.aways_available = aways_available
+
+        if condition_features is None:
+            self.condition_features = {}
+            self.condition_features['Current_filter'] = features.Current_filter()
+            self.condition_features['Mounted_filter'] = features.Mounted_filters()
+            self.condition_features['Sun_moon_alts'] = features.Sun_moon_alts()
+            self.condition_features['Current_mjd'] = features.Current_mjd()
+        if survey_features is None:
+            self.survey_features = {}
+            self.survey_features['Last_observation'] = features.Last_observation()
+            self.survey_features['N_obs_all'] = features.N_obs_count(filtername=None)
+            self.survey_features['N_obs'] = features.N_obs_count(filtername=filtername,
+                                                                 tag=tag)
+
+        super(Goal_Strict_filter_basis_function, self).__init__(survey_features=self.survey_features,
+                                                           condition_features=self.condition_features)
+
+    def filter_change_bonus(self, time):
+
+        lag_min = self.time_lag_min
+        lag_max = self.time_lag_max
+
+        a = 1. / (lag_max - lag_min)
+        b = -a * lag_min
+
+        bonus = a * time + b
+        # How far behind we are with respect to proportion?
+        nobs = self.survey_features['N_obs'].feature
+        nobs_all = self.survey_features['N_obs_all'].feature
+        goal = self.proportion
+        need = 1. - nobs / nobs_all + goal if nobs_all > 0 else 1. + goal
```

Comment:
The single line `if` `else` statement on this line, and those that follow, are fine, but I had to read them a couple of times to fully grasp what was happening. It is totally fine if they stay as they are.